### PR TITLE
[sdk] Error refactor

### LINKED ISSUE #238: [sdk] Error Refactor
In general, the sdk is using a single error type, which has many other errors embedded within it. This can make it difficult to parse meaningful context from the errors themselves. Instead, we should move to a pattern more similar to the one used by the router where the `message` contains very specific information and the `context` contains everything needed to debug *that* error only.

Heres a sample error I get when submitting a transaction fails:

```
{
  message: 'Error submitting transaction',
  context: {
    method: 'Contract::prepare',
    methodId: 'BzIxGGFeTV+RH//Ropocig/4851',
    txError: {
      message: 'cannot estimate gas; transaction may fail or may require manual gas limit (error={"reason":"cannot estimate gas; transaction may fail or may require manual gas limit","code":"UNPREDICTABLE_GAS_LIMIT","method":"estimateGas","transaction":{"from":"0x9e3b1BDafB95c0D4ab3e067F81A620fedA718dcd","maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x9502f900"},"maxFeePerGas":{"type":"BigNumber","hex":"0x9502f910"},"to":"0xD22dB9ad317A09384a86f310c973CDc350Bb510E","value":{"type":"BigNumber","hex":"0x00"},"data":"0x63405b9300000000000000000000000080623b8534d855479ff4621cabe65d451a81bdfe0000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd0000000000000000000000003b66db1b41b37510851cb03ec0be3377dce3f4220000000000000000000000008a1cad3703e0beae0e0237369b4fcd04228d16820000000000000000000000009ac2c46d7acc21c881154d57c0dc1c55a31391980000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd0000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000050000000000000000000000000000000000000000000000000000000000000004c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a47056d540e7b266de6be533ee06cef03f33dd994120daedfeacb83014512b91244a000000000000000000000000000000000000000000000000000000173f873b6b00000000000000000000000000000000000000000000000000000000612077210000000000000000000000000000000000000000000000000000000000000220000000000000000000000000000000000000000000000000000000000000024000000000000000000000000000000000000000000000000000000000000004c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000026000000000000000000000000000000000000000000000000000000000000000200000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd0000000000000000000000003b66db1b41b37510851cb03ec0be3377dce3f42200000000000000000000000000000000000000000000000000000000000000050000000000000000000000008a1cad3703e0beae0e0237369b4fcd04228d1682000000000000000000000000000000000000000000000000000000173f873b6b00000000000000000000000000000000000000000000000000000000000000040000000000000000000000009ac2c46d7acc21c881154d57c0dc1c55a3139198000000000000000000000000000000000000000000000000000000173c8d6fbd0000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd56d540e7b266de6be533ee06cef03f33dd994120daedfeacb83014512b91244a0000000000000000000000000000000000000000000000000000000061207721c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a47000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000220000000000000000000000000d22db9ad317a09384a86f310c973cdc350bb510e00000000000000000000000080623b8534d855479ff4621cabe65d451a81bdfe00000000000000000000000000000000000000000000000000000000611dc60d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000041fa83a234346b1cba74eef0f05bf99449bd887131022c2f91702382c998bc209b333b5f2f03a3076e68fa66d259e2086982d28ee8cd1bd395b61d24cb84f9d0091b00000000000000000000000000000000000000000000000000000000000000","type":2,"accessList":null}}, tx={"data":"0x63405b9300000000000000000000000080623b8534d855479ff4621cabe65d451a81bdfe0000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd0000000000000000000000003b66db1b41b37510851cb03ec0be3377dce3f4220000000000000000000000008a1cad3703e0beae0e0237369b4fcd04228d16820000000000000000000000009ac2c46d7acc21c881154d57c0dc1c55a31391980000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd0000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000050000000000000000000000000000000000000000000000000000000000000004c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a47056d540e7b266de6be533ee06cef03f33dd994120daedfeacb83014512b91244a000000000000000000000000000000000000000000000000000000173f873b6b00000000000000000000000000000000000000000000000000000000612077210000000000000000000000000000000000000000000000000000000000000220000000000000000000000000000000000000000000000000000000000000024000000000000000000000000000000000000000000000000000000000000004c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000026000000000000000000000000000000000000000000000000000000000000000200000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd0000000000000000000000003b66db1b41b37510851cb03ec0be3377dce3f42200000000000000000000000000000000000000000000000000000000000000050000000000000000000000008a1cad3703e0beae0e0237369b4fcd04228d1682000000000000000000000000000000000000000000000000000000173f873b6b00000000000000000000000000000000000000000000000000000000000000040000000000000000000000009ac2c46d7acc21c881154d57c0dc1c55a3139198000000000000000000000000000000000000000000000000000000173c8d6fbd0000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd56d540e7b266de6be533ee06cef03f33dd994120daedfeacb83014512b91244a0000000000000000000000000000000000000000000000000000000061207721c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a47000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000220000000000000000000000000d22db9ad317a09384a86f310c973cdc350bb510e00000000000000000000000080623b8534d855479ff4621cabe65d451a81bdfe00000000000000000000000000000000000000000000000000000000611dc60d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000041fa83a234346b1cba74eef0f05bf99449bd887131022c2f91702382c998bc209b333b5f2f03a3076e68fa66d259e2086982d28ee8cd1bd395b61d24cb84f9d0091b00000000000000000000000000000000000000000000000000000000000000","to":{},"from":"0x9e3b1BDafB95c0D4ab3e067F81A620fedA718dcd","value":{"type":"BigNumber","hex":"0x00"},"type":2,"maxFeePerGas":{"type":"BigNumber","hex":"0x9502f910"},"maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x9502f900"},"nonce":{},"gasLimit":{},"chainId":{}}, code=UNPREDICTABLE_GAS_LIMIT, version=abstract-signer/5.4.1)',
      type: 'Error',
      context: {},
      stack: 'Error: cannot estimate gas; transaction may fail or may require manual gas limit (error={"reason":"cannot estimate gas; transaction may fail or may require manual gas limit","code":"UNPREDICTABLE_GAS_LIMIT","method":"estimateGas","transaction":{"from":"0x9e3b1BDafB95c0D4ab3e067F81A620fedA718dcd","maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x9502f900"},"maxFeePerGas":{"type":"BigNumber","hex":"0x9502f910"},"to":"0xD22dB9ad317A09384a86f310c973CDc350Bb510E","value":{"type":"BigNumber","hex":"0x00"},"data":"0x63405b9300000000000000000000000080623b8534d855479ff4621cabe65d451a81bdfe0000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd0000000000000000000000003b66db1b41b37510851cb03ec0be3377dce3f4220000000000000000000000008a1cad3703e0beae0e0237369b4fcd04228d16820000000000000000000000009ac2c46d7acc21c881154d57c0dc1c55a31391980000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd0000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000050000000000000000000000000000000000000000000000000000000000000004c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a47056d540e7b266de6be533ee06cef03f33dd994120daedfeacb83014512b91244a000000000000000000000000000000000000000000000000000000173f873b6b00000000000000000000000000000000000000000000000000000000612077210000000000000000000000000000000000000000000000000000000000000220000000000000000000000000000000000000000000000000000000000000024000000000000000000000000000000000000000000000000000000000000004c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000026000000000000000000000000000000000000000000000000000000000000000200000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd0000000000000000000000003b66db1b41b37510851cb03ec0be3377dce3f42200000000000000000000000000000000000000000000000000000000000000050000000000000000000000008a1cad3703e0beae0e0237369b4fcd04228d1682000000000000000000000000000000000000000000000000000000173f873b6b00000000000000000000000000000000000000000000000000000000000000040000000000000000000000009ac2c46d7acc21c881154d57c0dc1c55a3139198000000000000000000000000000000000000000000000000000000173c8d6fbd0000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd56d540e7b266de6be533ee06cef03f33dd994120daedfeacb83014512b91244a0000000000000000000000000000000000000000000000000000000061207721c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a47000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000220000000000000000000000000d22db9ad317a09384a86f310c973cdc350bb510e00000000000000000000000080623b8534d855479ff4621cabe65d451a81bdfe00000000000000000000000000000000000000000000000000000000611dc60d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000041fa83a234346b1cba74eef0f05bf99449bd887131022c2f91702382c998bc209b333b5f2f03a3076e68fa66d259e2086982d28ee8cd1bd395b61d24cb84f9d0091b00000000000000000000000000000000000000000000000000000000000000","type":2,"accessList":null}}, tx={"data":"0x63405b9300000000000000000000000080623b8534d855479ff4621cabe65d451a81bdfe0000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd0000000000000000000000003b66db1b41b37510851cb03ec0be3377dce3f4220000000000000000000000008a1cad3703e0beae0e0237369b4fcd04228d16820000000000000000000000009ac2c46d7acc21c881154d57c0dc1c55a31391980000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd0000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000050000000000000000000000000000000000000000000000000000000000000004c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a47056d540e7b266de6be533ee06cef03f33dd994120daedfeacb83014512b91244a000000000000000000000000000000000000000000000000000000173f873b6b00000000000000000000000000000000000000000000000000000000612077210000000000000000000000000000000000000000000000000000000000000220000000000000000000000000000000000000000000000000000000000000024000000000000000000000000000000000000000000000000000000000000004c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000026000000000000000000000000000000000000000000000000000000000000000200000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd0000000000000000000000003b66db1b41b37510851cb03ec0be3377dce3f42200000000000000000000000000000000000000000000000000000000000000050000000000000000000000008a1cad3703e0beae0e0237369b4fcd04228d1682000000000000000000000000000000000000000000000000000000173f873b6b00000000000000000000000000000000000000000000000000000000000000040000000000000000000000009ac2c46d7acc21c881154d57c0dc1c55a3139198000000000000000000000000000000000000000000000000000000173c8d6fbd0000000000000000000000009e3b1bdafb95c0d4ab3e067f81a620feda718dcd56d540e7b266de6be533ee06cef03f33dd994120daedfeacb83014512b91244a0000000000000000000000000000000000000000000000000000000061207721c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a47000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000220000000000000000000000000d22db9ad317a09384a86f310c973cdc350bb510e00000000000000000000000080623b8534d855479ff4621cabe65d451a81bdfe00000000000000000000000000000000000000000000000000000000611dc60d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000041fa83a234346b1cba74eef0f05bf99449bd887131022c2f91702382c998bc209b333b5f2f03a3076e68fa66d259e2086982d28ee8cd1bd395b61d24cb84f9d0091b00000000000000000000000000000000000000000000000000000000000000","to":{},"from":"0x9e3b1BDafB95c0D4ab3e067F81A620fedA718dcd","value":{"type":"BigNumber","hex":"0x00"},"type":2,"maxFeePerGas":{"type":"BigNumber","hex":"0x9502f910"},"maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x9502f900"},"nonce":{},"gasLimit":{},"chainId":{}}, code=UNPREDICTABLE_GAS_LIMIT, version=abstract-signer/5.4.1)\n' +
        '    at Logger.makeError (/Users/laynehaber/Developer/nxtp/node_modules/@ethersproject/logger/src.ts/index.ts:213:28)\n' +
        '    at Logger.throwError (/Users/laynehaber/Developer/nxtp/node_modules/@ethersproject/logger/src.ts/index.ts:225:20)\n' +
        '    at /Users/laynehaber/Developer/nxtp/node_modules/@ethersproject/abstract-signer/src.ts/index.ts:301:31\n' +
        '    at runMicrotasks (<anonymous>)\n' +
        '    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n' +
        '    at async Promise.all (index 8)'
    }
  },
  type: 'TransactionManagerError',
  stack: 'Error: Error submitting transaction\n' +
    '    at /Users/laynehaber/Developer/nxtp/packages/sdk/src/transactionManager.ts:167:9\n' +
    '    at /Users/laynehaber/Developer/nxtp/node_modules/neverthrow/dist/index.cjs.js:230:102\n' +
    '    at runMicrotasks (<anonymous>)\n' +
    '    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n' +
    '    at async NxtpSdk.prepareTransfer (/Users/laynehaber/Developer/nxtp/packages/sdk/src/sdk.ts:976:24)\n' +
    '    at async /Users/laynehaber/Developer/nxtp/packages/integration/test/utils/sdkAgent.ts:267:9\n' +
    '    at async run (/Users/laynehaber/Developer/nxtp/node_modules/p-queue/dist/index.js:163:29)'
}
```

While this error does include all the stack traces, it is notably missing:
- `transactionId` being used
- `params` submitted to chain
- `message` does not indicate what transaction it was trying to send

### PR DESCRIPTION
**Pull request by jakekidd**
Refactor some of the erroring in SDK to be a bit more specific and descriptive. Perhaps not the most ideal data structure, but at least a step in the right direction and enough info to continue debugging for now. Also matches other modules' error structure a bit more.

Fixes #238

### CODE REVIEW

File: packages/sdk/src/error.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  PrepareParams,
+} from "@connext/nxtp-utils";
+import { providers } from "ethers";
+
+/**
+ * @classdesc Abstract error class thrown by the `TransactionManager` class.
+ */
+export class TransactionManagerError extends NxtpError {
+  static readonly type = TransactionManagerError.name;
+}
```
Comment (LayneHaber) [inline packages/sdk/src/error.ts:16]: Can you not declare this as `abstract`?

File: packages/sdk/src/error.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  constructor(
+    public readonly method: string,
+    public readonly methodId: string,
+    public readonly chainId: number,
+    public readonly context: {
+      transactionId?: string;
+      approveReceipt?: providers.TransactionReceipt;
+      assetId?: string;
+      amount?: string;
+      router?: string;
```
Comment (LayneHaber) [inline packages/sdk/src/error.ts:33]: Not sure if this is relevant to the error condition you're addressing here

File: packages/sdk/src/error.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ * @classdesc Defines the error thrown by the `TransactionManager` class when a transaction fails to be submitted.
+ */
+export class SubmitError extends TransactionManagerError {
+  static readonly type = SubmitError.name;
+
+  constructor(
+    public readonly method: string,
+    public readonly methodId: string,
+    public readonly chainId: number,
+    public readonly txserviceError: any,
```
Comment (LayneHaber) [inline packages/sdk/src/error.ts:59]: Can we type this as an `Error`?

File: packages/sdk/src/error.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    public readonly context: {
+      transactionId?: string;
+      approveReceipt?: providers.TransactionReceipt;
+      transactionData?: any;
+      params?: PrepareParams | FulfillParams | CancelParams | AuctionResponse;
+    } = {},
```
Comment (LayneHaber) [inline packages/sdk/src/error.ts:65]: can you do a `Partial` type here? a bit cleaner and has the same type signature

**Review by LayneHaber — CHANGES_REQUESTED** [review]
Overall good, would like the context to be less strictly typed in case there are additional things we want to add we are not constrained by the included typing. If there is something we want across all errors (i.e. `transactionId` or `user`) lets add those as constructor args rather than using the context as a catch all

File: packages/sdk/src/error.ts (hunk 1 above)
Comment (jakekidd) [reply packages/sdk/src/error.ts:16]: Lol yeah just didn't think to do so

File: packages/sdk/src/error.ts (hunk 2 above)
Comment (jakekidd) [reply packages/sdk/src/error.ts:33]: Me neither, I just added args bc I would rather err on side of too detailed

Comment (jakekidd) [reply packages/sdk/src/error.ts:33]: should I remove amount/router? assetId remove as well?

File: packages/sdk/src/error.ts (hunk 3 above)
Comment (jakekidd) [reply packages/sdk/src/error.ts:59]: Sure

File: packages/sdk/src/error.ts (hunk 4 above)
Comment (jakekidd) [reply packages/sdk/src/error.ts:65]: Partial of all of those?
`Partial<PrepareParams | FulfillParams | CancelParams | AuctionResponse>` ?

Comment (LayneHaber) [reply packages/sdk/src/error.ts:65]: `Partial<{ transactionId, approveReceipt, transactionData, params }>`

**Review by LayneHaber — APPROVED** [review]

**Pull request merged**